README.md
CircleCI Status
---------------

[![Circle CI](https://circleci.com/gh/zoff99/clewarecontrol/tree/build_with_circleCI.svg?style=svg)](https://circleci.com/gh/zoff99/clewarecontrol/tree/build_with_circleCI)


build instructions
------------------
Required libraries:
	libhidapi-dev

Since version 4.0 clewarecontrol uses this library to abstract the
interfacing with the kernel. That way it may work on other OSes and new(er)
kernel as well.

If you get errors, try replacing -lhidapi by -lhidraw-libusb or hidapi-hidraw.
If you still get errors, contact me.


Installation:
	make install


usage
-----
Please see the man-page for help on how to use it:
	make clewarecontrol

or run the program with the -h switch:
	clewarecontrol -h

Please also check the website for tips and help.


devices
-------
- temperature sensor
	clewarecontrol -rt

- humidity sensor
  This device needs to be "started" first, then you can read sensor readings
	clewarecontrol -ag -rh

- 2 channel AD converter (analog to digital)
  First you need to know what it range it works: 5v, 13v or 24v
  then you must select the channel to read from - the device has 2 channels
  after selecting the channel, you can keep reading
	clewarecontrol -ai x        where x is 0 or 1 (selects channel)
	clewarecontrol -ra y        where y is 0 voor 5v, 1 for 13v, etc.
 
- Ampel ("Switch1 (8)")
  The traffic light is, from clewarecontrol seen, a device with 3 switches in
  it that can be individiually switched on and off
	clewarecontrol -as x y      where x is 0, 1 or 2 (0=red, 2=green) and
                                    y is 0 or 1 (0 is off)

- IO16 ("Contact 00 device (48)")
  The IO16 is a 16 port I/O interface. You can configure each pin to be either
  input or output. This setting is stored in an eprom so that a pin cannot
  become accidently output while you expected an input.
  Set the direction with:
	clewarecontrol -ad x        where x is a hex value representing a 1
                                    or 0 selecting the direction
  After that you can set the output pins to 1 or 0 with:
	clewarecontrol -am x        where is a hex value

- Luminus
  This is in fact an AD converter with integrated hardware which measures the
  amount of light. It has 1 chnanel, so you don't need to select it (like the
  generic ADC). Reading values (in version 2.5) can be perfomed like his:
	clewarecontrol -ra 0
  This gives a value between 0 and 5. I (developer of this Linux program)
  need to buy a lux meter to scale this correctly.


- USB-cutter
  This is a device which is at its heart a switch like the other switches.
  To use it, you need to start it:
	clewarecontrol -ag
  then, you can toggle the USB status like this:
	clewarecontrol -as 0 1      this will "cut" the USB device (switching
	                            it off)
	clewarecontrol -as 0 0      this will enable the USB device again(!)
  indeed, the logic is inverted.


-mintrig / -maxtrig
-------------------
Run the program with -c 0 so that it keeps running forever (unless you want to
have it monitor the value for a shorter time).
Also start it with -F so that it runs in the background.
Then set the -mintrig or -maxtrig switches.
As the last parameter, set the sensor to monitor: -rh (humidity), -rt
(temperature), etc.
For example:
	clewarecontrol -c 0 -maxtrig 0.5 ./myscript.sh -ra 0
If the ADC tells us that the voltage is over 0.5 volts, then "myscript.sh" is
invoked. As a parameter, the script will get the measured value.


I would be delighted if you share your experiences with me!


python library
--------------
please see readme-python.txt for details on how to create and use the python
library.


perl library
--------------
please see readme-perl.txt for details on how to create and use the perl
library.


problems
--------
If you encounter any problems when using clewarecontrol with Cleware devices,
DO NOT contact Cleware but please contact me: folkert@vanheusden.com

Of course you can always write me when you have questions/suggestions/etc.
 folkert@vanheusden.com


FAQ
---
Q1. Values retrieved are invalid
A1. Occasionally the values retrieved are invalid. E.g. a temperature of 170
    degrees and such. In that case you can reset the device (might need to
    reset it multiple times!) and then it'll work

Q2. It does not compile on SuSE Enterprise Linux 10
A2. Insert in file USBaccessBasic.c at line 17 (that is BEFORE the line
    including usbdevice_fs.h!) the following line:
	#define __user

Q3. clewarecontrol -l doesn't list any devices
A3. The Cleware USB library expects the following files to exist:
	crw-rw-rw- 1 root root 180,  96 May  9  2005 /dev/usb/hiddev0
	crw-rw-rw- 1 root root 180,  97 May  9  2005 /dev/usb/hiddev1
	...
	crw-rw-rw- 1 root root 180, 110 May  9  2005 /dev/usb/hiddev14
	crw-rw-rw- 1 root root 180, 111 May  9  2005 /dev/usb/hiddev15

    Since version 1.2 you can use -p to select a different path, eg.:
	clewarecontrol -p /dev -l

    Make sure the user you're running clewarecontrol from has enought rights
    to access the devices. For starters, to see if it works at all, run it as
    root.

Q4. clewarecontrol still can't find any devices
A4. If that still does not give any output then please e-mail me the output of
    the following commands:
	uname -a
	lsusb
	lsmod
	strace clewarecontrol -l
	ps -deaf | grep udev
	find /dev -name hiddev\*

    Please also tell me what distribution (e.g. redhat/suse/ubuntu) and what
    version you're using


Regards,

Folkert van Heusden


debian/changelog
clewarecontrol (4.3-1) unstable; urgency=low

  * Initial release

 -- Donghoon Shin <dhs.shin@samsung.com>  Mon, 30 May 2016 11:49:42 +0900


tests/test_simple.py
import unittest

class ImportModule(unittest.TestCase):

    def test_import(self):
        try:
            import cleware
        except Exception as e:
            print(e)
            self.fail("unable to import the cleware module")

    def test_instantiate(self):
        try:
            import cleware
            c = cleware.CUSBaccess()        
        except Exception as e:
            print(e)
            self.fail("unable to instantiate a cleware access object")

    def test_enumerate(self):
        try:
            import cleware
            c = cleware.CUSBaccess()
            n_devices = c.OpenCleware()
        except Exception as e:
            print (e)
            self.fail("unable to enumerate the cleware devices")


USBaccess.h
// DLL class definitions for access to USB HID devices
//
// (C) 2001-2014 Copyright Cleware GmbH
// All rights reserved
//
// History:
// 05.01.2001	ws	Initial coding
// 17.07.2001	ws	cleanup interface
// 10.12.2001	ws	cleanup interface again, basic class to hide implementation details
// 13.12.2001	ws	introduced versionNumber and virtual destructor
// 23.05.2002	ws	added switch access
// ...
// 03.02.2003	ws	added switch version 10 
// 04.08.2003	ws	added humidity 
// 21.01.2004	ws	fixed some humidity problems 
//		   2004	ws	added contact + io16
// 05.02.2005	ws	added ADC08-Support 
// 25.05.13   ws	new controller support
// 15.05.2014	ws	4.3.0	added new controller support



// The following ifdef block is the standard way of creating macros which make exporting 
// from a DLL simpler. All files within this DLL are compiled with the USBACCESS_EXPORTS
// symbol defined on the command line. this symbol should not be defined on any project
// that uses this DLL. This way any other project whose source files include this file see 
// USBACCESS_API functions as being imported from a DLL, wheras this DLL sees symbols
// defined with this macro as being exported.


#ifndef __USBACCESS_H__
#define __USBACCESS_H__

#ifdef __APPLE__
#include "mac-hidapi/hidapi.h"
#else
#include <hidapi/hidapi.h>
#endif

typedef hid_device * HANDLE;

const int USBaccessVersion = 330 ;

class CUSBaccess {
	public:
		enum USBactions {		LEDs=0, EEwrite=1, EEread=2, Reset=3, KeepCalm=4, GetInfo=5, 
								StartMeasuring=6,		// USB-Humidity
								Configure=7,			// USB-IO16-V10, USB-Counter-V05
								RunPoint=10				// USB-Encoder
								} ;
		enum USBInfoType {		OnlineTime=1, OnlineCount=2, ManualTime=3, ManualCount=4 } ;
		enum LED_IDs {			LED_0=0, LED_1=1, LED_2=2, LED_3=3 } ;
		enum COUNTER_IDs {		COUNTER_0=0, COUNTER_1=1 } ;
		enum SWITCH_IDs {		SWITCH_0=0x10, SWITCH_1=0x11, SWITCH_2=0x12, SWITCH_3=0x13,
								SWITCH_4=0x14, SWITCH_5=0x15, SWITCH_6=0x16, SWITCH_7=0x17,
								SWITCH_8=0x18, SWITCH_9=0x19, SWITCH_10=0x1a, SWITCH_11=0x1b,
								SWITCH_12=0x1c, SWITCH_13=0x1d, SWITCH_14=0x1e, SWITCH_15=0x1f
								} ;
		enum USBtype_enum {		ILLEGAL_DEVICE=0,
								LED_DEVICE=0x01,
								POWER_DEVICE=0x02,
								WATCHDOG_DEVICE=0x05,
								AUTORESET_DEVICE=0x06,
								WATCHDOGXP_DEVICE=0x07,
								SWITCH1_DEVICE=0x08,
								SWITCH2_DEVICE=0x09, SWITCH3_DEVICE=0x0a, SWITCH4_DEVICE=0x0b,
								SWITCH5_DEVICE=0x0c, SWITCH6_DEVICE=0x0d, SWITCH7_DEVICE=0x0e, SWITCH8_DEVICE=0x0f,
								TEMPERATURE_DEVICE=0x10, 
								TEMPERATURE2_DEVICE=0x11,
								TEMPERATURE5_DEVICE=0x15, 
								HUMIDITY1_DEVICE=0x20,
								SWITCHX_DEVICE=0x28,		// new switch 3,4,8
								CONTACT00_DEVICE=0x30, CONTACT01_DEVICE=0x31, CONTACT02_DEVICE=0x32, CONTACT03_DEVICE=0x33, 
								CONTACT04_DEVICE=0x34, CONTACT05_DEVICE=0x35, CONTACT06_DEVICE=0x36, CONTACT07_DEVICE=0x37, 
								CONTACT08_DEVICE=0x38, CONTACT09_DEVICE=0x39, CONTACT10_DEVICE=0x3a, CONTACT11_DEVICE=0x3b, 
								CONTACT12_DEVICE=0x3c, CONTACT13_DEVICE=0x3d, CONTACT14_DEVICE=0x3e, CONTACT15_DEVICE=0x3f, 
								F4_DEVICE=0x40, 
								KEYC01_DEVICE=0x41, KEYC16_DEVICE=0x42,
								ADC0800_DEVICE=0x50, ADC0801_DEVICE=0x51, ADC0802_DEVICE=0x52, ADC0803_DEVICE=0x53, 
								COUNTER00_DEVICE=0x60, 
								ENCODER01_DEVICE=0x80,
								BUTTON_NODEVICE=0x1000
								} ;
	private:
		class CUSBaccessBasic *	X	;	// avoid export of internal USB variables

	public:
		CUSBaccess() ;
		virtual ~CUSBaccess() ;		// maybe used as base class

		virtual int			OpenCleware() ;			// returns number of found Cleware devices
		virtual int			CloseCleware() ;		// close all Cleware devices
		virtual int			Recover(int devNum) ;	// try to find disconnected devices, returns true if succeeded
		virtual HANDLE		GetHandle(int deviceNo) ;
		virtual int			GetValue(int deviceNo, unsigned char *buf, int bufsize) ;
		virtual int			SetValue(int deviceNo, unsigned char *buf, int bufsize) ;
		virtual int			SetLED(int deviceNo, enum LED_IDs Led, int value) ;	// value: 0=off 7=medium 15=highlight
		virtual int			SetSwitch(int deviceNo, enum SWITCH_IDs Switch, int On) ;	//	On: 0=off, 1=on
		virtual int			GetSwitch(int deviceNo, enum SWITCH_IDs Switch) ;			//	On: 0=off, 1=on, -1=error
		virtual int			GetSeqSwitch(int deviceNo, enum SWITCH_IDs Switch, int seqNum) ;		//	On: 0=off, 1=on, -1=error
		virtual int			GetSwitchConfig(int deviceNo, int *switchCount, int *buttonAvailable) ;
		virtual int			GetTemperature(int deviceNo, double *Temperature, int *timeID) ;
		virtual int			GetHumidity(int deviceNo, double *Humidity, int *timeID) ;
		virtual int			ResetDevice(int deviceNo) ;
		virtual int			StartDevice(int deviceNo) ;	
		virtual int			CalmWatchdog(int deviceNo, int minutes, int minutes2restart) ;
		virtual int			GetVersion(int deviceNo) ;
		virtual int			GetUSBType(int deviceNo) ;
		virtual int			GetSerialNumber(int deviceNo) ;
		virtual int			GetDLLVersion() { return USBaccessVersion ; }
		virtual int			GetManualOnCount(int deviceNo) ;		// returns how often switch is manually turned on
		virtual int			GetManualOnTime(int deviceNo) ;			// returns how long (seconds) switch is manually turned on
		virtual int			GetOnlineOnCount(int deviceNo) ;		// returns how often switch is turned on by USB command
		virtual int			GetOnlineOnTime(int deviceNo) ;			// returns how long (seconds) switch is turned on by USB command
		virtual int			GetMultiSwitch(int deviceNo, unsigned long int *mask, unsigned long int *value, int seqNumber) ;
		virtual int			SetMultiSwitch(int deviceNo, unsigned long int value) ;
		virtual int			SetMultiConfig(int deviceNo, unsigned long int directions) ;
		virtual int			GetCounter(int deviceNo, enum COUNTER_IDs counterID) ;	// COUNTER_IDs ununsed until now
		virtual int			SetCounter(int deviceNo, int counter, enum COUNTER_IDs counterID) ;	//  -1=error, COUNTER_IDs ununsed until now
		virtual int			SyncDevice(int deviceNo, unsigned long int mask) ;
		virtual int			GetHWversion(int deviceNo) ;	// return HWversion (0 for pre 2014 designed devices, 13 for new devices)
		virtual int			IOX(int deviceNo, int addr, int data) ;		// for internal use only, wrong usage may destroy device	
		virtual void		DebugWrite(char *s) ;
		virtual void		DebugWrite(char *f, int a1) ;
		virtual void		DebugWrite(char *f, int a1, int a2) ;
		virtual void		DebugWrite(char *f, int a1, int a2, int a3) ;
		virtual void		DebugWrite(char *f, int a1, int a2, int a3, int a4) ;
		virtual void		Sleep(int ms) { usleep(ms * 1000) ; }	// for Linux
	} ;

#endif // __USBACCESS_H__


USBaccessBasic.h
// Basic class definitions for access to USB HID devices
//
// (C) 2001 Copyright Cleware GmbH
// All rights reserved
//
// History:
// 05.01.01	ws	Initial coding
// 17.07.01	ws	cleanup interface
// 03.11.02	ws	small changes for Linux


#ifndef __USBACCESSBASIC_H__
#define __USBACCESSBASIC_H__

#ifdef __APPLE__
#include "mac-hidapi/hidapi.h"
#else
#include <hidapi/hidapi.h>
#endif

typedef hid_device * HANDLE;

enum USBtype_enum {	ILLEGAL_DEVICE=0,
						LED_DEVICE=0x01,
						POWER_DEVICE=0x02,
						DISPLAY_DEVICE=0x03,
						WATCHDOG_DEVICE=0x05,
						AUTORESET_DEVICE=0x06,
						WATCHDOGXP_DEVICE=0x07,
						SWITCH1_DEVICE=0x08,
						SWITCH2_DEVICE=0x09, SWITCH3_DEVICE=0x0a, SWITCH4_DEVICE=0x0b,
						SWITCH5_DEVICE=0x0c, SWITCH6_DEVICE=0x0d, SWITCH7_DEVICE=0x0e, SWITCH8_DEVICE=0x0f,
						TEMPERATURE_DEVICE=0x10,
						TEMPERATURE2_DEVICE=0x11,
						TEMPERATURE5_DEVICE=0x15, 
						HUMIDITY1_DEVICE=0x20,HUMIDITY2_DEVICE=0x21,
						SWITCHX_DEVICE=0x28,		// new switch 3,4,8
						// CONTACT1_DEVICE=0x30 
						CONTACT00_DEVICE=0x30, CONTACT01_DEVICE=0x31, CONTACT02_DEVICE=0x32, CONTACT03_DEVICE=0x33, 
						CONTACT04_DEVICE=0x34, CONTACT05_DEVICE=0x35, CONTACT06_DEVICE=0x36, CONTACT07_DEVICE=0x37, 
						CONTACT08_DEVICE=0x38, CONTACT09_DEVICE=0x39, CONTACT10_DEVICE=0x3a, CONTACT11_DEVICE=0x3b, 
						CONTACT12_DEVICE=0x3c, CONTACT13_DEVICE=0x3d, CONTACT14_DEVICE=0x3e, CONTACT15_DEVICE=0x3f, 
						F4_DEVICE=0x40, 
						KEYC01_DEVICE=0x41, KEYC16_DEVICE=0x42,MOUSE_DEVICE=0x43,
						ADC0800_DEVICE=0x50, ADC0801_DEVICE=0x51, ADC0802_DEVICE=0x52, ADC0803_DEVICE=0x53, 
						COUNTER00_DEVICE=0x60, 
						ENCODER01_DEVICE=0x80,
						BUTTON_NODEVICE=0x1000
						} ;
enum USBactions {		LEDs=0, EEwrite=1, EEread=2, Reset=3, KeepCalm=4, GetInfo=5, 
						StartMeasuring=6,		// USB-Humidity
						Configure=7,			// USB-IO16-V10, USB-Counter-V05
						Display=8,				// USB/Display
						RunPoint=10,			// USB-Encoder
						Programm=15				// Transfer new Firmware (internal use only)
						} ;

typedef struct {
	unsigned short v, p;
	hid_device *handle;
	int	gadgetVersionNo;
	enum USBtype_enum	gadgettype;
	int	SerialNumber;
	int	report_type;
	int	HWversion ;
	char	*hidpath;
	} SUSBdata;

extern int nr;

void cwInitCleware();
int	cwOpenCleware(const char *path);	// returns number of found Cleware devices
int cwRecover(int devNo);
void cwCloseCleware();
int	cwGetValue(int deviceNo, unsigned char *buf, int bufsize);
int	cwSetValue(int deviceNo, unsigned char *buf, int bufsize);
hid_device * cwGetHandle(int deviceNo);
int	cwGetVersion(int deviceNo);
int	cwGetSerialNumber(int deviceNo);
enum USBtype_enum	cwGetUSBType(int deviceNo);
int							cwValidSerNum(int SerialNumber, enum USBtype_enum devType) ;
void						cwDebugWrite(char *s) ;
void						cwDebugClose() ;
int cwGetHWversion(int deviceNo);
int							cwIOX(int deviceNo, int addr, int data) ;

#endif // __USBACCESS_H__


USBaccessBasic.cpp
// Basic class implementation for access to USB HID devices
//
// (C) 2001 Copyright Cleware GmbH
// All rights reserved
//
// Converted to HidApi by folkert van Heusden, mail@vanheusden.com
//
// History:
// 05.01.01	ws	Initial coding
// 01.11.01	ws	Linux coding
// 30.10.12	fvh	libusb version

#include <errno.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

#include "USBaccessBasic.h"

#define TIMEOUT 1000

const int maxHID = 256 ;
SUSBdata data[256] ;

int nr = 0;

void cwInitCleware()
{
	hid_init();
}

void cwCloseCleware()
{
	for(int h=0; h < maxHID; h++)
	{
		if (data[h].handle)
		{
			hid_close(data[h].handle);

			data[h].handle = NULL;
		}
	}

	hid_exit();
}

int	
cwIOX(int deviceNo, int addr, int datum) {	// return datum if ok, datum=-1=Read operation
	const int maxbufsize = 8 ;
	int bufsize = 6 ;
	unsigned char buf[maxbufsize] ;
	int ok = 1 ;
	if (deviceNo < 0 || deviceNo >= maxHID || data[deviceNo].handle == NULL)
		return-1 ;

	int devType = data[deviceNo].gadgettype ;
	int version = data[deviceNo].gadgetVersionNo ;
	int sixteenbit = (devType == TEMPERATURE2_DEVICE || devType == HUMIDITY1_DEVICE || devType == HUMIDITY2_DEVICE) ;

	if (datum >= 0) {		// -1 = Read command
		buf[0] = EEwrite ;
		if (sixteenbit) {
			buf[1] = addr >> 8 ;	// high byte 0
			buf[2] = addr ;
			buf[3] = datum ;
			cwSetValue(deviceNo, buf, 4) ;
			}
		else if (devType == CONTACT00_DEVICE && version > 6) {
			buf[1] = addr ;
			buf[2] = datum ;
			cwSetValue(deviceNo, buf, 5) ;
			}
		else if (devType == DISPLAY_DEVICE) {
			buf[1] = addr ;
			buf[2] = datum ;
			cwSetValue(deviceNo, buf, 5) ;
			}
		else if (devType == WATCHDOGXP_DEVICE || devType == SWITCHX_DEVICE) {
			buf[1] = addr ;
			buf[2] = datum ;
			cwSetValue(deviceNo, buf, 5) ;
			}
		else if (devType == ENCODER01_DEVICE) {
			buf[1] = addr ;
			buf[2] = datum ;
			cwSetValue(deviceNo, buf, 6) ;
			}
		else if (devType == ADC0800_DEVICE) {
			buf[1] = addr ;
			buf[2] = datum ;
			cwSetValue(deviceNo, buf, 3) ;
			}
		else if (devType == POWER_DEVICE) {
			buf[1] = addr ;
			buf[2] = datum ;
			cwSetValue(deviceNo, buf, 3) ;
			}
		else if (devType == KEYC16_DEVICE || devType == KEYC01_DEVICE) {
			buf[1] = addr ;
			buf[2] = datum ;
			cwSetValue(deviceNo, buf, 5) ;
			}
		else if (devType == MOUSE_DEVICE) {
			buf[1] = addr ;
			buf[2] = datum ;
			cwSetValue(deviceNo, buf, 5) ;
			}
		else {
			buf[1] = addr ;
			buf[2] = datum ;
			cwSetValue(deviceNo, buf, 3) ;
			}
		usleep(100*1000) ;
		}

	buf[0] = EEread ;
	if (sixteenbit) {
		buf[1] = 0 ;	// high byte 0
		buf[2] = addr ;
		buf[3] = 0 ;
		cwSetValue(deviceNo, buf, 4) ;
		bufsize = 7 ;
		}
	else if (devType == CONTACT00_DEVICE && version > 6) {
		buf[1] = addr ;
		buf[2] = 0 ;
		cwSetValue(deviceNo, buf, 5) ;
		}
	else if (devType == DISPLAY_DEVICE) {
		buf[1] = addr ;
		buf[2] = 0 ;
		cwSetValue(deviceNo, buf, 5) ;
		}
	else if (devType == WATCHDOGXP_DEVICE || devType == SWITCHX_DEVICE) {
		buf[1] = addr ;
		buf[2] = 0 ;
		cwSetValue(deviceNo, buf, 5) ;
		}
	else if (devType == KEYC16_DEVICE || devType == KEYC01_DEVICE) {
		buf[1] = addr ;
		buf[2] = 0 ;
		cwSetValue(deviceNo, buf, 5) ;
		bufsize = 8 ;
		}
	else if (devType == MOUSE_DEVICE) {
		buf[1] = addr ;
		buf[2] = 0 ;
		cwSetValue(deviceNo, buf, 5) ;
		bufsize = 4 ;
		}
	else if (devType == ADC0800_DEVICE) {
		buf[1] = addr ;
		buf[2] = 0 ;
		cwSetValue(deviceNo, buf, 3) ;
		bufsize = 4 ;
		}
	else if (devType == POWER_DEVICE) {
		buf[1] = addr ;
		buf[2] = 0 ;
		cwSetValue(deviceNo, buf, 3) ;
		bufsize = 3 ;
		}
	else if (devType == ENCODER01_DEVICE) {
		buf[1] = addr ;
		buf[2] = 0 ;
		cwSetValue(deviceNo, buf, 6) ;
		}
	else {
		buf[1] = addr ;
		buf[2] = 0 ;
		cwSetValue(deviceNo, buf, 3) ;
		}

	usleep(10*1000) ;
	ok = 40 ;
	int Xdata = -1 ;
	while (ok) {
		if (cwGetValue(deviceNo, buf, bufsize)) {
			if ((buf[0] & 0x80) == 0) {
				if (--ok == 0) {
					// MessageBox("GetValue still not valid", "Error") ;
					break ;
					}
				else {
					usleep(10*1000) ;
					continue ;
					}
				}
			int Xaddr = 0 ;
			if (bufsize == 3 || devType == MOUSE_DEVICE) {
				Xaddr = buf[1] ;
				Xdata = buf[2] ;
				}
			else if (bufsize == 4) {
				Xaddr = buf[2] ;
				Xdata = buf[3] ;
				}
			else {
				Xaddr = buf[4] ;
				Xdata = buf[5] ;
				}
			if (sixteenbit) {
				Xaddr = (Xaddr << 8) + buf[5] ;
				Xdata = buf[6] ;
				}
			if (Xaddr != addr) {
				if (--ok == 0) {
					// MessageBox("GetValue address error", "Error") ;
					break ;
					}
				else {
					usleep(10*1000) ;
					continue ;
					}
				}
			if (datum >= 0 && Xdata != datum) {
				if (--ok == 0) {
					// MessageBox("Write error", "Error") ;
					break ;
					}
				else {
					usleep(10*1000) ;
					continue ;
					}
				}
			break ;
			}
		else {
			if (--ok == 0) {
				// MessageBox("GetValue failed", "Error") ;
				break ;
				}
			else {
				usleep(10*1000) ;
				continue ;
				}
			}
		break ;		// read was ok
		}
	if (!ok)
		Xdata = -1 ;

	return Xdata ;
	}

// returns number of found Cleware devices
int cwOpenCleware(const char *path = NULL)
{
	int n = 0;
	struct hid_device_info *devs = NULL, *cur_dev = NULL;

	memset(&data, 0x00, sizeof data);

	cur_dev = devs = hid_enumerate(0x0d50, 0x0);

	while (cur_dev)
	{
		data[n].v = cur_dev->vendor_id;
		data[n].p = cur_dev->product_id;
		data[n].hidpath = strdup(cur_dev->path);

		data[n].gadgettype = (enum USBtype_enum)data[n].p;
		data[n].gadgetVersionNo = cur_dev -> release_number;

                char buffer[256] = { 0 };
                wcstombs(buffer, cur_dev->serial_number, wcslen(cur_dev->serial_number));

                data[n].SerialNumber = strtol(buffer, NULL, 16);
		data[n].report_type = 123; // HID_REPORT_ID_FIRST; // NOT USED *FVH*
		if (data[n].SerialNumber == 0x63813) {	// this is the next controller - get serial number directly
			data[n].HWversion = 13 ;
			data[n].SerialNumber = -1 ;
		}
		if (data[n].SerialNumber <= 0) {		// getting the Serial number failed, so get it directly!
			data[n].handle = hid_open_path(data[n].hidpath);
			int SerNum = 0 ;
			int addr = 0;
			for (addr=8 ; addr <= 14 ; addr++) {	// unicode byte 2 == 0
				int data = cwIOX(n, addr, -1) ;
				if (data >= '0' && data <= '9')
					SerNum = SerNum * 16 + data - '0' ;
				else if (data >= 'A' && data <= 'F')
					SerNum = SerNum * 16 + data - 'A' + 10 ;
				else {
					SerNum = -1 ;		// failed!
					break ;
				}
			}
			data[n].SerialNumber = SerNum ;
			hid_close(data[n].handle);
			data[n].handle = NULL;
		}

		n++;
		cur_dev = cur_dev->next;
	}

	hid_free_enumeration(devs);

	return n;
}

int cwOpenDevice(int index)
{
	if (data[index].handle)
		return 1;

	data[index].handle = hid_open_path(data[index].hidpath);

	char ok = data[index].handle != NULL;

	if (!ok)
		fprintf(stderr, "Failed to open %04x:%04x (%d)\n", data[index].v, data[index].p, data[index].SerialNumber);

	return ok;
}

int cwCloseDevice(int index)
{
	if (data[index].handle)
	{
		hid_close(data[index].handle);

		data[index].handle = NULL;
	}

	return 1;
}

int cwRecover(int devNo)
{
	if (data[devNo].handle == NULL)
		return 0;

	return 1;
}

void fail(const char *msg, hid_device *handle)
{
	const wchar_t *wstr = hid_error(handle);

	if (!wstr)
	{
		if (errno)
			printf("%s: %s\n", msg, strerror(errno));
		else
			printf("%s\n", msg);
	}
	else
	{
		int len = wcslen(wstr);
		char* ascii = new char[len + 1];

		wcstombs(ascii, wstr, len);

		printf("%s: %s\n", msg, ascii);

		delete [] ascii;
	}
}

int cwGetValue(int deviceNo, unsigned char *buf, int bufsize)
{
	if (!cwOpenDevice(deviceNo))
		return 0;

	if (data[deviceNo].handle == NULL)
		return 0;

	unsigned char lbuf[3] = { 0x00, (unsigned char)(nr++), 0x81 };
	if (hid_send_feature_report(data[deviceNo].handle, lbuf, sizeof lbuf) < 0)
	{
		fail("cwGetValue::hid_write() failed", data[deviceNo].handle);

                return 0;
        }

	if (hid_read(data[deviceNo].handle, buf, bufsize) < 0)
	{
		fail("cwGetValue::hid_read() failed", data[deviceNo].handle);

                return 0;
        }

	return 1;
}

int cwSetValue(int deviceNo, unsigned char *buf, int bufsize)
{
	if (!cwOpenDevice(deviceNo))
		return 0;

	if (data[deviceNo].handle == NULL)
		return 0;

	char *b = new char[bufsize + 1];
	memcpy(b + 1, buf, bufsize);
	b[0] = 0x00;

//#ifdef macos
#if 1
	if (hid_write(data[deviceNo].handle, (unsigned char *)b, bufsize + 1) < 0)
#else
	if (hid_send_feature_report(data[deviceNo].handle, (unsigned char *)b, bufsize + 1) < 0)
#endif
	{
		fail("cwSetValue::hid_write() failed", data[deviceNo].handle);

		delete [] b;

                return 0;
        }

	delete [] b;

	return 1;
}

hid_device * cwGetHandle(int deviceNo)
{ 
	if (!cwOpenDevice(deviceNo))
		return 0;

	return data[deviceNo].handle;
}

int cwGetVersion(int deviceNo)
{
	return data[deviceNo].gadgetVersionNo;
}

int cwGetSerialNumber(int deviceNo)
{ 
	return data[deviceNo].SerialNumber;
}

enum USBtype_enum cwGetUSBType(int deviceNo)
{ 
	return data[deviceNo].gadgettype;
}

int
cwValidSerNum(int SerialNumber, enum USBtype_enum devType) {
	static int outdated[] = {
		54,59,60,62,63,64,65,66,67,68,69,
		72,74,75,76,77,82,83,85,87,88,89,
		90,91,92,93,95,96,98,99,
		100,101,102,103,105,106,107,108,109,110,
		113,116,117,119,120,122,124,125,126,128,
		131,132,135,139,140,142,143,145,147,148,149,
		150,151,152,153,154,155,156,157,160,161,162,163,168,169,
		170,171,172,173,174,175,176,180,184,187,188,189,190,191,192,193,195,197,198,
		201,203,204, 205, 206, 219,220,221,260,272,273,274,275,276,278,279,
		280,281,416,
		5002,5003,5004,5005,5006,5007,5008,5010,
		5011,5012,5013,5014,5015,5016,
		5017,5018,5019,5020,5021,5022,5023,5024,
		5025,5026,5028,5029,5032,5033,5034,
		5035,5036,5041,5043,5044,5046,5049,5050,
		5052,5053,5055,5057,5071,5073,5076,5089,5091,5101,5102,5103,
		5104,5106,5109,5114,5116,5117,5118,5119,
		5120,5121,5122,5147,5163,5164,7502,7503,
		7504,7505,7511,7513,8192,8193,8194,
		8503,8504,8505,8506,8507,8508,8509,8510,8511,8512,8513
		} ;

	static int outdatedSwitches[] = {		// double numbers!!
		510,511,513,514,517,518,520,532			//	Switches
//		, 5314	// test
		} ;

	int rval = 1 ;

	int size = sizeof(outdated) / sizeof(int) ;
	int *pt = &(outdated[0]) ;
	int i ;

	for (i=0 ; i < size ; i++, pt++) {
		if (SerialNumber == *pt) {
			rval = 0 ;
			break ;
			}
		}
	if (rval == 1 && devType == SWITCH1_DEVICE) {
		size = sizeof(outdatedSwitches) / sizeof(int) ;
		pt = &(outdatedSwitches[0]) ;
		for (i=0 ; i < size ; i++, pt++) {
			if (SerialNumber == *pt) {
				rval = 0 ;
				break ;
				}
			}
		}

	return rval ; 
	}

void
cwDebugWrite(char *s) { 
	fputs(s, stderr) ;
	}

void
cwDebugClose() { 
	}

int
cwGetHWversion(int deviceNo) {                  // return current
        int rval = 0 ;

        if (deviceNo < 0 || deviceNo >= maxHID || data[deviceNo].handle == NULL)
                rval = -1 ;
        else
                rval = data[deviceNo].HWversion ;

        return rval ;
        }
